build/Shared/EqualityUtility.cs

```diff
@@ -36,6 +36,83 @@ internal static class EqualityUtility
                 .SequenceEqual(other.OrderBy(keySelector, orderComparer), sequenceComparer);
         }
 
+        /// <summary>
+        /// Compares two collections for equality, ordered according to the specified key and optional comparer. Handles null values gracefully.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the list</typeparam>
+        /// <typeparam name="TKey">The type of the sorting key</typeparam>
+        /// <param name="self">This list</param>
+        /// <param name="other">The other list</param>
+        /// <param name="keySelector">The function to extract the key from each item in the list</param>
+        /// <param name="orderComparer">An optional comparer for comparing keys</param>
+        /// <param name="sequenceComparer">An optional comparer for sequences</param>
+        internal static bool OrderedEquals<TSource, TKey>(this ICollection<TSource> self, ICollection<TSource> other, Func<TSource, TKey> keySelector, IComparer<TKey> orderComparer = null, IEqualityComparer<TSource> sequenceComparer = null)
+        {
+            Debug.Assert(orderComparer != null || typeof(TKey) != typeof(string), "Argument " + "orderComparer" + " must be provided if " + "TKey" + " is a string.");
+            Debug.Assert(sequenceComparer != null || typeof(TSource) != typeof(string), "Argument " + "sequenceComparer" + " must be provided if " + "TSource" + " is a string.");
+
+            bool identityEquals;
+            if (TryIdentityEquals(self, other, out identityEquals))
+            {
+                return identityEquals;
+            }
+
+            if (self.Count != other.Count)
+            {
+                return false;
+            }
+
+            if (self.Count == 0)
+            {
+                return true;
+            }
+
+            return self
+                .OrderBy(keySelector, orderComparer)
+                .SequenceEqual(other.OrderBy(keySelector, orderComparer), sequenceComparer);
+        }
+
+        /// <summary>
+        /// Compares two lists for equality, ordered according to the specified key and optional comparer. Handles null values gracefully.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the list</typeparam>
+        /// <typeparam name="TKey">The type of the sorting key</typeparam>
+        /// <param name="self">This list</param>
+        /// <param name="other">The other list</param>
+        /// <param name="keySelector">The function to extract the key from each item in the list</param>
+        /// <param name="orderComparer">An optional comparer for comparing keys</param>
+        /// <param name="sequenceComparer">An optional comparer for sequences</param>
+        internal static bool OrderedEquals<TSource, TKey>(this IList<TSource> self, IList<TSource> other, Func<TSource, TKey> keySelector, IComparer<TKey> orderComparer = null, IEqualityComparer<TSource> sequenceComparer = null)
```

Comment:
Since `IList<T>` extends `ICollection<T>`, is this overload needed?

Reply:
The `IList<T>` overload lets us deal with the case where both lists have a `Count` of 1 without calling `OrderBy`/`SequenceEqual` (line 106).

Reply:
@zivkan 
Do you still have a concern if we merge as it's now?